pkg/clusterclientstore/cache/cluster.go

```diff
@@ -18,34 +18,41 @@ const (
 func NewCluster(
 	name,
 	checksum string,
-	client kubernetes.Interface,
 	config *rest.Config,
 	informerFactory kubeinformers.SharedInformerFactory,
+	buildClient clientBuilderFunc,
 	cacheSyncCb func(),
 ) *cluster {
 	return &cluster{
 		state:           StateNotReady,
 		name:            name,
 		checksum:        checksum,
-		client:          client,
 		config:          config,
 		informerFactory: informerFactory,
+		buildClient:     buildClient,
 		cacheSyncCb:     cacheSyncCb,
 		stopCh:          make(chan struct{}),
+
+		clients: make(map[string]kubernetes.Interface),
 	}
 }
 
+type clientBuilderFunc func(string, string, *rest.Config) (kubernetes.Interface, error)
+
 type cluster struct {
 	name string
 
 	stateMut sync.RWMutex
 	state    string
 
+	clientsMut sync.RWMutex
+	clients    map[string]kubernetes.Interface
```

Comment:
Clients might be sync.Map, in this case clientsMut would be gone.

Reply:
I didn't know about sync.Map, thanks for the suggestion. If I read the docs correctly, though, I could only not *store* a new client if it's already there, but there is no way to only *generate* a client if there isn't one. It's a bit nitpicky though. What do you think?

Reply:
I see, you are right, I must have missed the point. In this case, using a mutex makes more sense. A quick suggestion: consider using sync.Mutex instead as it seems you have no one-writer-multiple-readers kind of pattern, and the aforementioned structure is much more lightweight and faster. Another way to solve this is to use sync.Once(func()), which can guarantee a single-time initialization. All options are quite equivalent in terms of following the best practices. Feel free to choose any :-)